src/identity_committer.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{
+    app::{Hash, TreeState},
+    contracts::Contracts,
+    database::Database,
+};
+use eyre::eyre;
+use std::sync::{atomic::Ordering, Arc};
+use tokio::{
+    select,
+    sync::{mpsc, mpsc::error::TrySendError, RwLock},
+    task::JoinHandle,
+};
+use tracing::{debug, error, info, instrument};
+
+struct RunningInstance {
+    #[allow(dead_code)]
+    handle:          JoinHandle<eyre::Result<()>>,
+    wake_up_sender:  mpsc::Sender<()>,
+    shutdown_sender: mpsc::Sender<()>,
+}
+
+impl RunningInstance {
+    fn wake_up(&self) -> eyre::Result<()> {
+        // We're using a 1-element channel for wake-up notifications. It is safe to
+        // ignore a full channel, because that means the committer is already scheduled
+        // to wake up and will process all requests inserted in the database.
+        match self.wake_up_sender.try_send(()) {
+            Ok(_) => {
+                debug!("Scheduled a committer job.");
+                Ok(())
+            }
+            Err(TrySendError::Full(_)) => {
+                debug!("Committer job already scheduled.");
+                Ok(())
+            }
+            Err(TrySendError::Closed(_)) => Err(eyre!("Committer thread terminated unexpectedly.")),
+        }
+    }
+
+    async fn shutdown(self) -> eyre::Result<()> {
+        info!("Sending a shutdown signal to the committer.");
+        // Ignoring errors here, since we have two options: either the channel is full,
+        // which is impossible, since this is the only use, and this method takes
+        // ownership, or the channel is closed, which means the committer thread is
+        // already dead.
+        self.shutdown_sender.send(()).await.ok();
+        info!("Awaiting committer shutdown.");
+        self.handle.await?
+    }
+}
+
+pub struct IdentityCommitter {
+    instance:   RwLock<Option<RunningInstance>>,
+    database:   Arc<Database>,
+    contracts:  Arc<Contracts>,
+    tree_state: Arc<TreeState>,
+}
+
+impl IdentityCommitter {
+    pub fn new(
+        database: Arc<Database>,
+        contracts: Arc<Contracts>,
+        tree_state: Arc<TreeState>,
+    ) -> Self {
+        Self {
+            instance: RwLock::new(None),
+            database,
+            contracts,
+            tree_state,
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    pub async fn start(&self) {
+        let mut instance = self.instance.write().await;
+        if instance.is_some() {
+            info!("Identity committer already running");
+            return;
+        }
+        let (shutdown_sender, mut shutdown_receiver) = mpsc::channel(1);
+        let (wake_up_sender, mut wake_up_receiver) = mpsc::channel(1);
+        let database = self.database.clone();
+        let tree_state = self.tree_state.clone();
+        let contracts = self.contracts.clone();
+        let handle = tokio::spawn(async move {
```

Comment:
This basically creates a closure taking the variables cloned above. In the future when this gets more complex we probably want a explicit struct for the context like so:

```rust
struct Task {
  database: // ...
  // ..
}

impl Task {
   async fn run(mut self) { ... }
}
```

And then launch it by 

```
let task = Task { ... };
let handler = tokio::spawn(task.run());
```